README.md
# cimir: Interface to CIMIS <a href='https://github.com/mkoohafkan/cimir'><img src='man/figures/logo.png' align="right" height="139" /></a>

<!-- badges: start -->
[![R build status](https://github.com/mkoohafkan/cimir/workflows/R-CMD-check/badge.svg)](https://github.com/mkoohafkan/cimir/actions)
![CRAN Release](https://www.r-pkg.org/badges/version-last-release/cimir)
<!-- badges: end -->

This package provides an R interface to the 
[California Irrigation Management Information System](https://cimis.water.ca.gov/)
(CIMIS) [Web API](http://et.water.ca.gov/Home/Index). In order to use 
this package, you will need to 
[create a CIMIS account](https://cimis.water.ca.gov/Auth/Register.aspx) 
and [request a web services AppKey](https://et.water.ca.gov/Home/Register/). 
Browse the documentation at 
[mkoohafkan.github.io/cimir](https://mkoohafkan.github.io/cimir/).


NEWS.md
# cimir 0.4-1 (2021-02-17)

This patch provides better handling of queries
that return empty record sets.

# cimir 0.4-0 (2020-01-22)

This release includes the following changes:

- References to defunct functions were removed.
- Minor changes to internals to meet new requirements in latest versions 
  of `tidyr` and `rlang` packages.
- Fixed bug in function `cimis_to_datetime()` regarding timezone 
  specification.
- Fixed bug in function `cimis_split_query()` where sub-queries 
  sometimes exceeded the specified length.
- New function `cimis_flags()` provides information on CIMIS Quality 
  Control flags.
- Added a new vignette describing helper functions.
- Error is returned if an empty or NULL API key is passed to 
  `set_key()`.
- Informative error message is returned when the CIMIS API itself 
  rejects a request.


# cimir 0.3-0 (2019-06-10)

This release includes the following changes:

- **(Breaking change)** Deprecated functions are now defunct.
- Package import "Rcurl" replaced with "curl" (>= 3.3).
- Data items not listed in the CIMIS Web API documentation were 
  identified and added to output of `cimis_items()`.
- Function `cimis_to_datetime()` now explicitly uses Etc/GMT+8 
  timezone to ensure match with standard timezone used by CIMIS (PST).
- New helper function `cimis_split_query()` splits a long-duration 
  query into multiple smaller-duration queries.
- New helper function `cimis_compass_to_degrees()` returns the middle 
  azimuths of a vector of compass direction labels.
- New helper function `cimis_format_location()` formats station 
  latitudes and longitudes as either decimal degrees (numeric) or 
  Hours Minutes Seconds.
- Text encoding of Web API results are set to "UTF-8", fixing potential 
  formatting issues.
- HTTP errors now include the message text provided by CIMIS to 
  describe the issue.

# cimir 0.2-0 (2019-03-14)

This release includes the following changes:

- **(Warning)** Function names have been updated to follow best 
  (or at least better) practices for avoiding
  naming conflicts and improving clarity. Old function names
  have been deprecated.
- HTTPS is now used to access the CIMIS web API.
- The URL formatting has bee updated to reflect a change in
  how the CIMIS web API formats the prioritizeSCS argument.
- The quickstart vignette has been updated to use the new function
  names.

# cimir 0.1-0 (2019-01-04)

This is the first submission of this package to CRAN. 

R/query.r
base.url = "https://et.water.ca.gov/api"


# Default CIMIS query items
default.items = c("day-asce-eto", "day-precip", "day-sol-rad-avg",
  "day-vap-pres-avg", "day-air-tmp-max", "day-air-tmp-min",
  "day-air-tmp-avg", "day-rel-hum-max", "day-rel-hum-min",
  "day-rel-hum-avg", "day-dew-pnt", "day-wind-spd-avg",
  "day-wind-run", "day-soil-tmp-avg")

#' dataitems
#'
#' A tibble of data items and their names, classes, and providers.
#' @docType data
#' @keywords internal
"dataitems"


#' CIMIS Data Items
#'
#' List CIMIS data items.
#'
#' @param type The type of data item, i.e. `"Daily"` or `"Hourly"`.
#' @return a dataframe of data items.
#'
#' @examples
#' cimis_items()
#'
#' @importFrom stringr str_to_title
#' @importFrom dplyr filter
#' @export
cimis_items = function(type = c("Daily", "Hourly")) {
  type = match.arg(str_to_title(type), c("Daily", "Hourly"), TRUE)
  filter(dataitems, .data$Class %in% type)
}

#' CIMIS Data Flags
#'
#' List CIMIS data quality control flags.
#'
#' @param type The type of data flag, i.e. `"Severe"` or
#'   `"Informative"`.
#' @param period The Time period that data was collected, i.e.,
#'   "Current" or "Former" (pre-1995).
#' @return a dataframe of data flags.
#'
#' @seealso [CIMIS Data Overview - Quality Control](https://cimis.water.ca.gov/Resources.aspx)
#'
#' @examples
#' cimis_flags()
#' cimis_flags("Informative")
#' cimis_flags("Severe", period = "Former")
#'
#' @importFrom stringr str_to_title
#' @importFrom dplyr filter
#' @export
cimis_flags = function(type = c("Severe", "Informative"),
  period = "Current") {
  type = match.arg(str_to_title(type), c("Severe", "Informative"), TRUE)
  period = match.arg(str_to_title(period), c("Current", "Former"), TRUE)
  filter(dataflags, .data$Class %in% type, .data$Period %in% period)
}

#' Query CIMIS Data
#'
#' Query CIMIS data using the Web API.
#'
#' @param targets geographies or weather stations of interest. This
#'   parameter may specify one or many stations, zip codes,
#'   coordinates, or street addresses; however, you are not allowed to
#'   mix values from different categories. This means the targets
#'   parameter must contain only stations, only zip codes, only
#'   coordinates, or only street addresses. You will receive an error
#'   if you attempt to mix different category types. The formats are
#'   accepted:
#'   - A comma delimited list of WSN station numbers
#'   - A comma delimited list of California zip codes
#'   - A semicolon delimited list of decimal - degree coordinates
#'   - A semicolon delimited list of street addresses
#' @param items specifies one or more comma-delimited data elements to
#'   include in your response. See `data_items()` for a complete list
#'   of possible data element values. Default: day-asce-eto,
#'   day-precip, day-sol-rad-avg, day-vap-pres-avg, day-air-tmp-max,
#'   day-air-tmp-min, day-air-tmp-avg, day-rel-hum-max,
#'   day-rel-hum-min, day-rel-hum-avg, day-dew-pnt, day-wind-spd-avg,
#'   day-wind-run, day-soil-tmp-avg.
#' @param start.date Specifies the start date. The data format is
#'   `"yyyy-mm-dd"`.
#' @param end.date Specifies the end date. The data format is
#'   `"yyyy-mm-dd"`.
#' @param measure.unit The unit of measure may be either `"E"` for
#'   English units or `"M"` for metric units. The value of this
#'   parameter will affect data values in the response. For
#'   example, designating English units will result in temperature
#'   values being returned in Fahrenheit rather than Celsius.
#' @param prioritize.SCS This parameter is relevant only when the
#'   targets parameter contains zip code(s). If `TRUE`, the Spatial
#'   CIMIS System (SCS) will be used as the preferred data provider.
#' @return A `tibble` object.
#'
#' @examples
#' if(is_key_set()) {
#'   cimis_data(targets = 170, start.date = Sys.Date() - 4,
#'     end.date = Sys.Date() - 1)
#' }
#'
#' @importFrom glue glue
#' @importFrom stringr str_c str_to_upper
#' @export
cimis_data = function(targets, start.date, end.date, items,
  measure.unit = c("E", "M"), prioritize.SCS = TRUE) {
  if (any(is.na(suppressWarnings(as.numeric(targets))))) {
    target.sep = ";"
  } else {
    target.sep = ","
  }
  if (missing(items))
    items = default.items
  measure.unit = match.arg(str_to_upper(measure.unit), c("E", "M"),
    FALSE)
  prioritize.SCS = ifelse(prioritize.SCS, "Y", "N")
  start.date = as.Date(start.date)
  end.date = as.Date(end.date)
  # query
  result = basic_query(
    glue("{base.url}/data?",
      "appKey={authenv$appkey}", "&",
      "targets={str_c(targets, collapse = target.sep)}", "&",
      "startDate={start.date}", "&",
      "endDate={end.date}", "&",
      "dataItems={str_c(items, collapse = ',')}", "&",
      "unitOfMeasure={measure.unit}", "&",
      "prioritizeSCS={prioritize.SCS}"
    )
  )
  bind_records(result)
}


#' Query CIMIS Station Metadata
#'
#' Query CIMIS station metadata.
#'
#' @param station The station ID. If missing, metadata for all stations
#'   is returned.
#' @return A `tibble` object.
#'
#' @examples
#' if(is_key_set()) {
#'   cimis_station()
#'   cimis_zipcode()
#'   cimis_spatial_zipcode()
#' }
#' @importFrom purrr map map_dfr
#' @importFrom glue glue
#' @importFrom dplyr as_tibble bind_rows
#' @export
cimis_station = function(station) {
  if (missing(station)) {
    url = glue("{base.url}/station")
  } else {
    url = glue("{base.url}/station/{station}")
  }
  result = map(url, basic_query)
  map_dfr(result, function(s) map_dfr(s$Stations, as_tibble))
}


#' @rdname cimis_station
#'
#' @param zipcode The (spatial) zip code. If missing, metadata for all
#'   stations is returned.
#'
#' @importFrom purrr map_dfr
#' @importFrom glue glue
#' @importFrom dplyr as_tibble bind_rows
#' @export
cimis_spatial_zipcode = function(zipcode) {
  if (missing(zipcode)) {
    url = glue("{base.url}/spatialzipcode")
  } else {
    url = glue("{base.url}/spatialzipcode/{zipcode}")
  }
  result = map(url, basic_query)
  map_dfr(result, function(s) map_dfr(s$ZipCodes, as_tibble))
}


#' @rdname cimis_station
#'
#' @importFrom purrr map_dfr
#' @importFrom glue glue
#' @importFrom dplyr as_tibble bind_rows
#' @export
cimis_zipcode = function(zipcode) {
  if (missing(zipcode)) {
    url = glue("{base.url}/stationzipcode")
  } else {
    url = glue("{base.url}/stationzipcode/{zipcode}")
  }
  result = map(url, basic_query)
  map_dfr(result, function(s) map_dfr(s$ZipCodes, as_tibble))
}


#' Basic Query
#'
#' Helper function for CIMIS query handling.
#'
#' @param url The query URL.
#' @return The parsed JSON string, as a list.
#'
#' @importFrom curl curl_fetch_memory parse_headers
#' @importFrom jsonlite fromJSON
#' @importFrom stringr str_replace_all
#' @keywords internal
basic_query = function(url) {
  if (!is_key_set())
    stop("No API key available. Specify key with \"set_key()\".")
  result = curl_fetch_memory(url, handle = cimir_handle())
  if (result$status_code != 200L)
    stop("CIMIS query failed with status ",
      parse_headers(result$headers)[1], "\n",
      parse(text = rawToChar(result$content)), "\n",
      "URL request: ", result$url,
      call. = FALSE)
  value = rawToChar(result$content)
  Encoding(value) = "UTF-8"
  # check if request was rejected
  if (str_detect(value, "Request Rejected")) {
    stop("The CIMIS API returned an error. ",
      "Check that your API key is correct.",
      "\n", "CIMIS error message:", "\n",
      value, call. = FALSE)
  }
  fromJSON(
    str_replace_all(value, c(":null" = ":[null]",
      ":\\[\\]" = ":\\[\\[\\]\\]")),
    simplifyDataFrame = FALSE)
  # can also replace [] with [[]] or [null]
}

#' cimir curl handle
#'
#' Get the handle for curl URL handling in cimir.
#'
#' @importFrom curl new_handle handle_setopt handle_setheaders
#' @keywords internal
cimir_handle = function() {
  h = new_handle()
  handle_setopt(h, connecttimeout = getOption("cimir.timeout"))
  handle_setheaders(h, Accept = "application/json")
  h
}


R/defunct.r
#' @title Defunct functions in package \pkg{cimir}.
#' @description The functions listed below were deprecated in the previous 
#'   release and are now defunct. Alternative functions with similar
#'   functionality are mentioned. Help pages for defunct functions are
#'   available at \code{help("cimir-defunct")}.
#' @name cimir-defunct
#' @keywords internal
NULL

#' @rdname cimir-defunct
#' @section \code{get_data}:
#' For \code{get_data}, use \code{\link{cimis_data}}.
#'
#' @export
get_data = function(targets, start.date, end.date, items,
  measure.unit = c("E", "M"), prioritize.SCS = TRUE) {
  .Defunct("cimis_data")
  cimis_data(targets, start.date, end.date, items,
  measure.unit, prioritize.SCS)
}

#' @rdname cimir-defunct
#' @section \code{get_station}:
#' For \code{get_station}, use \code{\link{cimis_station}}.
#'
#' @export
get_station = function(station) {
  .Defunct("cimis_station")
  cimis_station(station)
}

#' @rdname cimir-defunct
#' @section \code{get_station_spatial_zipcode}:
#' For \code{get_station_spatial_zipcode}, use \code{\link{cimis_spatial_zipcode}}.
#'
#' @export
get_station_spatial_zipcode = function(zipcode) {
  .Defunct("cimis_spatial_zipcode")
  cimis_spatial_zipcode(zipcode)
}

#' @rdname cimir-defunct
#' @section \code{get_station_zipcode}:
#' For \code{get_station_zipcode}, use \code{\link{cimis_zipcode}}.
#'
#' @export
get_station_zipcode = function(zipcode) {
  .Defunct("cimis_zipcode")
  cimis_zipcode(zipcode)
}


#' @rdname cimir-defunct
#' @section \code{to_datetime}:
#' For \code{to_datetime}, use \code{\link{cimis_to_datetime}}.
#'
#' @export
to_datetime = function(d) {
  .Defunct("cimis_to_datetime")
  cimis_to_datetime(d)
}

#' @rdname cimir-defunct
#' @section \code{data_items}:
#' For \code{data_items}, use \code{\link{cimis_items}}.
#'
#' @export
data_items = function(type = c("Daily", "Hourly")) {
  .Defunct("cimis_items")
  cimis_items(type)
}


R/package.r
#' cimir: Interface to CIMIS
#'
#' This package provides an R interface to the 
#' [California Irrigation Management Information System](https://cimis.water.ca.gov)
#' (CIMIS) [Web API](https://et.water.ca.gov). In order to use 
#' this package, you will need to 
#' [create a CIMIS account](https://cimis.water.ca.gov/Auth/Register.aspx) 
#' and request a web services AppKey. 
#'
#' @section Package options:
#'
#' cimir uses the following [options()] to configure behavior:
#'
#' \itemize{
#'   \item `cimir.appkey`: The CIMIS AppKey to use for queries.
#'   \item `cimir.timeout`: The maximum time to wait for a response 
#'     from the CIMIS Web API.
#' }
#' Alternatively, the CIMIS App Key can be saved to an environment
#' variable `CIMIS_APPKEY`.
#'
#' @docType package
#' @name cimir
"_PACKAGE"



.onLoad = function(libname, pkgname) {
  # set up environment for authentication storage
  authenv = new.env(parent = getNamespace(pkgname))
  assign("authenv", authenv, envir = getNamespace(pkgname))
}

.onAttach = function(libname, pkgname) {
  # check for existing AppKey
  if (!is.null(options()[["cimir.appkey"]]) &&
      nchar(options()[["cimir.appkey"]]) > 0L) {
    set_key(options()[["cimir.appkey"]])
    packageStartupMessage("Using existing App Key from ",
      'option "cimir.appkey".')
  } else if (nchar(Sys.getenv("CIMIS_APPKEY")) > 0L) {
    set_key(Sys.getenv("CIMIS_APPKEY"))
    packageStartupMessage("Using existing App Key from ",
      '"CIMIS_APPKEY" environment variable.')
  } else {
    remove_key()
    packageStartupMessage('Set your CIMIS App Key with "set_key()".')
  }
  if (is.null(options()[["cimir.timeout"]])) {
    options(cimir.timeout = 30)
  }
}

.onDetach = function(libname) {
  #delete authentication information
  remove_key()
}


R/deprecated.r
#' @title Deprecated functions in package \pkg{cimir}.
#' @description The functions listed below are deprecated and will be defunct in
#'   the near future. When possible, alternative functions with similar
#'   functionality are also mentioned. Help pages for deprecated functions are
#'   available at \code{help("-deprecated")}.
#' @name cimir-deprecated
#' @keywords internal
NULL

#' @rdname cimir-deprecated
#' @section \code{get_data}:
#' For \code{get_data}, use \code{\link{cimis_data}}.
#'
#' @export
get_data = function(targets, start.date, end.date, items,
  measure.unit = c("E", "M"), prioritize.SCS = TRUE) {
  .Deprecated("cimis_data")
  cimis_data(targets, start.date, end.date, items,
  measure.unit, prioritize.SCS)
}

#' @rdname cimir-deprecated
#' @section \code{get_station}:
#' For \code{get_station}, use \code{\link{cimis_station}}.
#'
#' @export
get_station = function(station) {
  warning('Function "get_station" is deprecated. ',
    'Use "cimis_station" instead.', call. = FALSE)
  cimis_station(station)
}

#' @rdname cimir-deprecated
#' @section \code{get_station_spatial_zipcode}:
#' For \code{get_station_spatial_zipcode}, use \code{\link{cimis_spatial_zipcode}}.
#'
#' @export
get_station_spatial_zipcode = function(zipcode) {
  .Deprecated("cimis_spatial_zipcode")
  cimis_spatial_zipcode(zipcode)
}

#' @rdname cimir-deprecated
#' @section \code{get_station_zipcode}:
#' For \code{get_station_zipcode}, use \code{\link{cimis_zipcode}}.
#'
#' @export
get_station_zipcode = function(zipcode) {
  .Deprecated("cimis_zipcode")
  cimis_zipcode(zipcode)
}


#' @rdname cimir-deprecated
#' @section \code{to_datetime}:
#' For \code{to_datetime}, use \code{\link{cimis_to_datetime}}.
#'
#' @export
to_datetime = function(d) {
  .Deprecated("cimis_to_datetime")
  cimis_to_datetime(d)
}

#' @rdname cimir-deprecated
#' @section \code{data_items}:
#' For \code{data_items}, use \code{\link{cimis_items}}.
#'
#' @export
data_items = function(type = c("Daily", "Hourly")) {
  .Deprecated("cimis_items")
  cimis_items(type)
}


R/util.r
cimis.tz = "Etc/GMT+8"

empty.record = tibble(
  Date = as.Date(character(0)),
  Hour = character(0),
  Julian = integer(0),
  Station = character(0),
  Standard = character(0),
  ZipCodes = character(0),
  Scope = character(0),
  Item = character(0),
  Value = character(0),
  Qc = character(0)
)

#' Quick Fix to as_tibble
#'
#' Handle empty lists when coercing to tibble. See
#' [tibble issue 851](https://github.com/tidyverse/tibble/issues/851).
#'
#' @param d An object to coerce to a tibble.
#' @return a tibble.
#' @importFrom purrr modify_if
#' @keywords internal
as_tibble_fix = function(d) {
  as_tibble(modify_if(d, ~ identical(.x, list()),
    ~ list(NULL)))
}

#' To Datetime
#'
#' Collapse The Date and Hour columns to a single DateTime Column.
#'
#' @param d A data frame of CIMIS data results.
#' @return The data frame, with a new `"Datetime"` column replacing
#'   the `"Date"` and `"Hour"` columns.
#'
#' @details According to the
#'   [CIMIS Report FAQs](https://cimis.water.ca.gov/Default.aspx),
#'   all CIMIS data is based on Pacific Standard Time (PST).
#'
#' @examples
#' if(is_key_set()) {
#'   d = cimis_data(targets = 170, start.date = Sys.Date() - 4,
#'     end.date = Sys.Date() - 1, items = "hly-air-tmp")
#'   cimis_to_datetime(d)
#' }
#' @importFrom dplyr select mutate if_else rename
#' @importFrom stringr str_c
#' @export
cimis_to_datetime = function(d) {
  if (!("Hour" %in% names(d)))
    d = mutate(d, Hour = "0000")
  rename(select(mutate(d,
    Hour = if_else(is.na(.data$Hour), "0000", .data$Hour),
    Date = as.POSIXct(str_c(.data$Date, " ", .data$Hour),
      format = "%Y-%m-%d %H%M", tz = cimis.tz)),
    -.data$Hour
  ), Datetime = .data$Date)
}


#' Record to Data Frame
#'
#' Convert a single record, containing one or more data items, to a to
#' a single data frame.
#'
#' @param record A single CIMIS record, in list format.
#' @return A data frame. The column `"Item"` identifies the data item.
#'
#' @importFrom tidyr unnest
#' @importFrom dplyr mutate bind_rows setdiff as_tibble
#' @importFrom purrr map
#' @importFrom rlang .data
#' @keywords internal
record_to_df = function(record) {
  if (identical(record, list())) {
    return(empty.record)
  }
  fixed = c("Date", "Hour", "Julian", "Station", "Standard",
    "ZipCodes", "Scope")
  data.names = setdiff(names(record), fixed)
  other.names = setdiff(names(record), data.names)
  unnest(mutate(as_tibble(record[other.names]),
    Date = as.Date(.data$Date),
    Julian = as.integer(.data$Julian),
    Data = list(bind_rows(map(record[data.names], as_tibble),
      .id = "Item"))
  ), cols = c(.data$Data))
}


#' Bind Records
#'
#' Bind CIMIS records into a single data frame. This function
#'   is used internally.
#'
#' @param result CIMIS query results.
#' @return A data frame.
#'
#' @importFrom tidyr unnest
#' @importFrom purrr map_dfr
#' @importFrom dplyr mutate bind_rows as_tibble case_when
#'   across matches
#' @importFrom rlang .data
#' @keywords internal
bind_records = function(result) {
  mutate(unnest(mutate(
    map_dfr(result[[c("Data", "Providers")]], as_tibble),
    Records = map(.data$Records, record_to_df)),
    cols = c(.data$Records)), across(matches("Value"), as.numeric))
}

#' Split CIMIS Query
#'
#' Split a large CIMIS query into multiple smaller queries based on a
#' time interval.
#'
#' @inheritParams cimis_data
#' @param max.records The maximum number of records returned by a
#'   query. The default value is the the maximum data limit allowed by
#'   the CIMIS Web API (1,750 records).
#' @return A data frame with columns "targets", "start.date",
#'   "end.date", and "items".
#'
#' @details Queries are not split by `targets` or `items`, i.e. each
#'   resulting query will include all targets and items.
#'
#' @examples
#' cimis_split_query(170, "2000-01-01", "2010-12-31", "day-air-tmp-avg")
#' cimis_split_query(c(149, 170), "2018-01-01", "2018-12-31",
#'   c("day-air-tmp-avg", "hly-air-tmp", "hly-rel-hum"))
#'
#' @importFrom dplyr tibble n mutate bind_rows
#' @export
cimis_split_query = function(targets, start.date, end.date, items,
  max.records = 1750L) {
  hourly.items = intersect(items, cimis_items("Hourly")[["Data Item"]])
  daily.items = intersect(items, cimis_items("Daily")[["Data Item"]])
  if (length(hourly.items) > 0L) {
    hourly.ranges = mutate(date_seq(start.date, end.date, max.records,
      24 * length(targets) * length(hourly.items)),
      items = rep(list(hourly.items), n()))
  } else {
    hourly.ranges = NULL
  }
  if (length(daily.items) > 0L) {
    daily.ranges = mutate(date_seq(start.date, end.date, max.records,
      length(targets) * length(daily.items)),
      items = rep(list(daily.items), n()))
  } else {
    daily.ranges = NULL
  }
  mutate(bind_rows(daily.ranges, hourly.ranges),
    targets = rep(list(targets), n()))
}

#' @importFrom dplyr tibble
#' @importFrom utils head tail
#' @keywords internal
date_seq = function(start.date, end.date, max.length, multiplier) {
  start.date = as.Date(start.date)
  end.date = as.Date(end.date)
  num.records = as.numeric(end.date - start.date) * multiplier
  if (num.records < max.length) {
    tibble(start.date = start.date, end.date = end.date)
  } else {
    num.queries = as.integer(ceiling(num.records / max.length))
    seq.start = seq(start.date, end.date, length.out = num.queries + 1)
    starts = head(seq.start, -1)
    ends = c(head(tail(seq.start, -1), -1) - 1, tail(seq.start, 1))
    tibble(start.date = starts, end.date = ends)
  }
}


#' Compass Direction To Degrees
#'
#' Convert the Compass direction labels to degrees.
#'
#' @param x A vector of compass directions, i.e. the data item labels
#'  "DayWindNnw", "DayWindSse", etc. Recognized directions are
#'   North-northeast (NNE), East-northeast (ENE), East-southeast (ESE),
#'   South-southeast (SSE), South-southwest (SSW), West-southwest (WSW),
#'   West-northwest (WNW), and North-northwest (NNW).
#'
#' @return A numeric vector of degrees corresponding to the middle
#'   azimuth of the corresponding compass direction.
#'
#' @examples
#' cimis_compass_to_degrees("day-wind-nne")
#' cimis_compass_to_degrees(c("SSE", "SSW", "wsw", "Wnw", "nnw"))
#'
#' @seealso [cimis_degrees_to_compass()]
#'
#' @importFrom dplyr case_when
#' @importFrom stringr str_to_upper str_detect
#' @export
cimis_compass_to_degrees = function(x) {
  x = str_to_upper(x)
  res = case_when(
    str_detect(x, "NNE$") ~ 22.5,
    str_detect(x, "ENE$") ~ 67.5,
    str_detect(x, "ESE$") ~ 112.5,
    str_detect(x, "SSE$") ~ 157.5,
    str_detect(x, "SSW$") ~ 202.5,
    str_detect(x, "WSW$") ~ 247.5,
    str_detect(x, "WNW$") ~ 292.5,
    str_detect(x, "NNW$") ~ 337.5,
    TRUE ~ NA_real_
  )
  if (any(is.na(res)))
    stop("Unrecognized values in arugment \"x\".")
  res
}

#' Degrees to Compass Direction
#'
#' Convert decimal degrees to Compass direction.
#'
#' @param x A vector of directions in decimal degrees.
#' @return A factor vector of compass directions.
#'
#' @details Degrees are labeled with their corresponding
#'   Primary InterCardinal compass direction, following the
#'   convention of the CIMIS daily wind data items.
#'
#' @examples
#' cimis_degrees_to_compass(c(30, 83, 120, 140, 190, 240, 300, 330))
#' cimis_degrees_to_compass(cimis_compass_to_degrees(c("NNE", "ENE",
#'   "ESE", "SSE", "SSW", "WSW", "WNW", "NNW")))
#'
#' @seealso [cimis_compass_to_degrees()]
#' @export
cimis_degrees_to_compass = function(x) {
  breaks = c(0, 45, 90, 135, 180, 225, 270, 315, 360)
  labels = c("NNE", "ENE", "ESE", "SSE", "SSW", "WSW", "WNW", "NNW")
  cut(x, breaks, labels, include.lowest = TRUE)
}


#' Format CIMIS Station Location
#'
#' Format the latitude and longitude of station in
#'   Decimal Degrees (DD) or Hour Minutes Seconds (HMS).
#'
#' @inheritParams cimis_to_datetime
#' @param format The format to use, either Decimal Degrees (`"DD"`)
#'   or Hour Minutes Seconds (`"HMS"`).
#'
#' @return The data frame, with a new `"Latitude"` and `"Longitude"`
#'   columns replacing the `"HmsLatitude"` and `"HmsLongitude"`
#'   columns.
#'
#' @examples
#' if(is_key_set()) {
#'   d = cimis_station(170)
#'   cimis_format_location(d, "DD")
#'   cimis_format_location(d, "HMS")
#' }
#'
#' @importFrom dplyr mutate_at rename
#' @importFrom stringr str_split str_replace
#' @export
cimis_format_location = function(d, format = c("DD", "HMS")) {
  format = match.arg(str_to_upper(format), c("DD", "HMS"))
  if (format == "HMS") {
    fun = function(x)
      str_replace(str_split(x, " / ", simplify = TRUE)[, 1], "^-", "")
    } else {
      fun = function(x)
        as.numeric(str_split(x, " / ", simplify = TRUE)[, 2])
      }
  rename(
    mutate_at(d, c("HmsLatitude", "HmsLongitude"), fun),
    Latitude = .data$HmsLatitude, Longitude = .data$HmsLongitude
  )
}
